docs/troubleshooting.rst

```diff
@@ -277,6 +277,127 @@ the log level of the command to the DEBUG level and the ``-X`` option shows stac
 
    $ inmanta -vvv export -X
 
+
+Compilation fails
+=================
+
+In rare situations, the compiler might fail with a ``List modified after freeze`` or an
+``Optional variable accessed that has no value`` error, even thought the model is syntactically correct. The following
+sections describe why this error occurs and what can be done to make the compilation succeed.
+
+Reason for compilation failure
+------------------------------
+
+When the compiler runs, it cannot know upfront how many elements will be added to a relationship. At some stages of the
+compilation process the compiler has to guess which relations are completely populated in order to be able to continue
+the compilation process. Heuristics are being used to determine the correct order in which relationships can be
+considered completely populated. In most situation these heuristics work well, but in rare situations the compiler
+makes an incorrect decision and considers a relationship to be complete while it isn't. In those situation the compiler
+crashes with one of the following exception:
+
+* ``List modified after freeze``: This error occurs when a relationship with an upper arity larger than one was
+  considered complete too soon.
+* ``Optional variable accessed that has no value``: This error occurs when a ``[0:1]`` relationship was considered
+  complete too soon.
+
+The following sections provide information on how this issue can be resolved.
+
+Relationship precedence policy
+------------------------------
+
+.. warning::
+
+    The relationship precedence policy feature is currently experimental
+
+The above-mentioned problem can be resolved by defining a *relation precedence policy* in the ``project.yml``
+file of an Inmanta project. This policy consists of a list of rules. Each rule defining the order in which two
+relationships should be considered complete with respect to each other. By providing this policy, it's possible to
+guide the compiler in making the correct decisions that lead to a successful compilation.
+
+Example: Consider the following ``project.yml`` file.
+
+.. code-block:: text
+    :linenos:
+
+    name: quickstart
+    modulepath: libs
+    downloadpath: libs
+    repo: https://github.com/inmanta/
+    description: A quickstart project that installs a drupal website.
+    relation_precedence_policy:
+      - "A::Submod_A.relation before B::Submod_B.other_relation"
```

Comment:
These names do not seem to follow our conventions (I'm not even sure if they're correct syntax), which I would try to avoid in our docs. Also, the names `Submod_{A,B}` seem to indicate they're modules while they're actually entities.